### PR TITLE
Collision detection maintenance

### PR DESCRIPTION
**Pull request by jngrad**
Description of changes:
- convert the Cython file `collision_detection.pyx` to a Python file
- reduce the complexity of the Python code by hiding implementation details in the C++ ScriptInterface class
- delegate the Python-to-C++ type conversion logic to the `AutoParameters` class instance
- bugfix: a runtime warning would print at random during collision due to undefined behavior
- fix broken error message: `glue_to_surface` and `bind_at_point_of_collision` require `VIRTUAL_SITES_RELATIVE`
- test all error messages
- calling `collision_detection.set_params()` with invalid arguments no longer leaves the system in an indeterminate state

### CODE REVIEW

**Comment by jngrad** [comment]
Note: this PR is only meant to make the code more maintainable for the 4.2.0 release. It also lays the groundwork for a possible overhaul in a future minor release, e.g. 4.3.0 (#4483).

**Comment by jngrad** [comment]
ping

**Review by reinaual — DISMISSED** [review]
I guess you are right with #4483...

File: src/script_interface/collision_detection/CollisionDetection.hpp [hunk 1]
```diff
@@ -59,18 +97,97 @@ class CollisionDetection : public AutoParameters<CollisionDetection> {
...
+    auto const mode = get_value<std::string>(params, "mode");
+    auto const expected_params = get_param_names(mode);
+    auto const expected_param_names =
+        std::set<std::string>{expected_params.begin(), expected_params.end()};
+    std::set<std::string> input_parameter_names = {};
+    for (auto const &kv : params) {
+      auto const param_name = get_value<std::string>(kv.first);
+      if (expected_param_names.count(param_name) == 0) {
+        // throw if parameter is unknown
+        static_cast<void>(get_parameter(param_name));
```
Comment (reinaual) [inline src/script_interface/collision_detection/CollisionDetection.hpp:172]: this is probably violating the SRP...

**Review by reinaual — APPROVED** [review]

**Pull request merged**